src/components/tools/listing/Listing.js

```diff
@@ -0,0 +1,154 @@
+    };
+
+    // Handles a request to chagne the number of rows per page.
+    const handleChangeRowsPerPage = (newPageSize) => {
+        setRowsPerPage(parseInt(newPageSize, 10));
+        setSelected([]);
+        setPage(0);
+    };
+
+    // Determines whether or not a tool is already selected.
+    const isSelected = (toolId) => selected.includes(toolId);
+
+    // Adds zero or more tools to the list of selected tools.
+    const select = (toolIds) => {
+        setSelected([...new Set([...selected, ...toolIds])]);
+    };
+
+    // Removes zero or more tools from the list of selected tools.
+    const deselect = (toolIds) => {
+        setSelected(selected.filter((id) => !toolIds.includes(id)));
+    };
+
+    // Toggles the selection of an individual tool.
+    const toggleSelection = (toolId) => {
+        isSelected(toolId) ? deselect([toolId]) : select([toolId]);
+    };
+
+    // Selects all of the tools in an index range.
+    const rangeSelect = (start, end, targetId) => {
+        // Ensure that the start index comes before the end index.
+        if (start > end) {
+            [start, end] = [end, start];
+        }
+
+        // Ignore the case where the user hasn't selected a starting index.
+        if (start < 0) {
+            return;
+        }
+
+        // Get the tool IDs for the range that the user selected.
+        const rangeIds = [];
+        for (let i = start; i <= end; i++) {
+            rangeIds.push(data?.tools[i].id);
+        }
+
+        // Toggle the selection based on the last tool clicked.
+        isSelected(targetId) ? deselect(rangeIds) : select(rangeIds);
+    };
+
+    // Handles a click on a single tool.
+    const handleClick = (event, id, index) => {
+        if (event.shiftKey) {
+            rangeSelect(lastSelectedIndex, index, id);
+        } else {
+            toggleSelection(id);
+        }
+        setLastSelectedIndex(index);
+    };
+
+    // Handles a request to select all tools.
+    const handleSelectAllClick = (event) => {
+        console.warn("handling a select all event");
+        setSelected(
+            event.target.checked && !selected.length
+                ? data?.tools?.map((tool) => tool.id) || []
+                : []
+        );
+    };
+
+    return (
+        <>
+            <TableView
+                baseId={baseId}
+                error={error}
+                handleClick={handleClick}
+                handleRequestSort={handleRequestSort}
+                handleSelectAllClick={handleSelectAllClick}
+                listing={data}
+                loading={isFetching}
+                order={order}
+                orderBy={orderBy}
+                selected={selected}
+            />
+            {data && data.total > 0 && (
+                <DEPagination
+                    baseId={baseId}
+                    onChange={handleChangePage}
+                    onPageSizeChange={handleChangeRowsPerPage}
+                    page={page + 1}
```

Comment:
Not really related to this PR since we have this everywhere, but I realized in `handlePageChange` we have `setPage(newPage - 1)` and here we have `page + 1` which makes me think we can just use `page` everywhere?